kerastuner/engine/tuner.py

```diff
@@ -229,27 +237,32 @@ def run_trial(self, trial, hp, fit_args, fit_kwargs):
             # Get model; this will reset the Keras session
             if not self.tune_new_entries:
                 hp = hp.copy()
-            model = self._build_model(hp)
-            self._compile_model(model)
 
-            # Start execution
-            execution = execution_module.Execution(
-                execution_id=execution_id,
-                trial_id=trial.trial_id,
-                max_epochs=max_epochs,
-                max_steps=max_steps,
-                base_directory=trial.directory)
-            trial.executions.append(execution)
-            self.on_execution_begin(trial, execution, model)
-
-            # During model `fit`,
-            # the patched callbacks call
-            # `self.on_epoch_begin`, `self.on_epoch_end`,
-            # `self.on_batch_begin`, `self.on_batch_end`.
-            fit_kwargs['callbacks'] = self._inject_callbacks(
-                original_callbacks, trial, execution)
-            model.fit(*fit_args, **fit_kwargs)
-            self.on_execution_end(trial, execution, model)
+            with tuner_utils.maybe_distribute(self.distribution_strategy):
```

Comment:
Can we do this in `build_model` and `compile_model` instead? `run_trial` should be overridable by users, so it should contain as little logic as possible.

Reply:
Done